TrigScint/src/TrigScint/Firmware/clusterproducer_sw.cxx

```diff
@@ -0,0 +1,77 @@
+#include <stdio.h>
+
+#include <iostream>
+
+#include "TrigScint/Firmware/clusterproducer.h"
+#include "TrigScint/Firmware/objdef.h"
+
+std::unique_ptr<Cluster[]> clusterproducer_sw(Hit inHit[NHITS]) {
```

Comment:
What you want to use is `std::array<Cluster, NCLUS>` since you know the length of the array at compile time. This allows you to pass around C-style, fixed-length arrays in C++ without needing to have bare pointers or any `new`.

Reply:
This is what I'm imagining: https://godbolt.org/z/zbYzxKPzc

You'll have to manually enforce the shortening of `NCLUS` to `NTRK` but I think that is beneficial since it points out to the user a design of the firmware that we are intentionally emulating in the software.